src/main/resources/org/csanchez/jenkins/plugins/kubernetes/pipeline/KubernetesDeclarativeAgentScript.groovy

```diff
@@ -0,0 +1,64 @@
+/*
+ * The MIT License
+ *
+ * Copyright (c) 2016, CloudBees, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.csanchez.jenkins.plugins.kubernetes.pipeline
+
+import hudson.model.Result
+import org.jenkinsci.plugins.pipeline.modeldefinition.SyntheticStageNames
+import org.jenkinsci.plugins.pipeline.modeldefinition.agent.DeclarativeAgentScript
+import org.jenkinsci.plugins.workflow.cps.CpsScript
+
+
+public class KubernetesDeclarativeAgentScript extends DeclarativeAgentScript<KubernetesDeclarativeAgent> {
+    public KubernetesDeclarativeAgentScript(CpsScript s, KubernetesDeclarativeAgent a) {
+        super(s, a)
+    }
+
+    @Override
+    public Closure run(Closure body) {
+        return {
+            try {
+                script.podTemplate(describable.asArgs) {
+                    script.node(describable.label) {
+                        if (describable.isDoCheckout() && describable.hasScmContext(script)) {
+                            if (!describable.inStage) {
+                                script.stage(SyntheticStageNames.checkout()) {
+                                    script.checkout script.scm
+                                }
+                            } else {
+                                // No stage when we're in a nested stage already
+                                script.checkout script.scm
+                            }
+                        }
+                        script.container(describable.containerTemplate.name) {
```

Comment:
I am not sure that wrapping every call in `container(describable.containerTemplate.name){ }` is such a good idea.

All commands are executed inside the one single container template that can be specified, while the pod will _at least_ contain a 'jnlp' container in addition. For users to execute scripts in the jnlp container, they need to specify it:

```
step {
   container('jnlp') {
      sh 'echo run on the slave'
   }
}
```.

Also, the current example contains an(other) explicit `container`, see https://github.com/jenkinsci/kubernetes-plugin/blob/132c66c3f47e40982a9af822d9d5b49eba3f6862/examples/declarative.groovy#L17.

I would vote for removing the `container` here, making the default "execution location" the jnlp container (like it is for non-declarative pipeline). That will also enable us to add support for multiple containerTemplates in declarative pipeline.

Or was it a conscious decision to keep the declarative syntax simpler and therefore less powerful?

@carlossg what do you think?

Reply:
Just saw https://github.com/jenkinsci/kubernetes-plugin/pull/127#issuecomment-278701279, i.e. multiple containerTemplates may not make sense..

Reply:
I believe I know the answer (from following the above thread) but does the current declarative support for this plugin support podTemplate inheritance? Such that you can specify X number of containers you want in the POD? (I think the answer is no).

Reply:
@marvinthepa  @carlossg Just so I am sure I understand because I cannot find this explicitly mentioned in the docs: If in a declarative pipeline you do not wrap an "sh" build step in a container, it is always executed in the container with name "jnlp"?